Copied changes related w/ product API to this branch

BrainDias opened the pull request.

BrainDias committed:
Copied changes related w/ product API to this branch from GreetingService_repository_and_tests

ahlinist commented on src/main/java/com/example/demo/controllers/ProductController.java:
```diff
@@ -0,0 +1,61 @@
+@RequiredArgsConstructor
+@RequestMapping("/products")
+public class ProductController {
+    private final UserService userService;
```
blank line (see oracle code conventions)

ahlinist commented on src/main/java/com/example/demo/controllers/ProductController.java:
```diff
@@ -0,0 +1,61 @@
+import lombok.SneakyThrows;
+import org.springframework.http.HttpStatus;
+import org.springframework.security.core.annotation.AuthenticationPrincipal;
+import org.springframework.web.bind.annotation.*;
```
try not to use asterisk imports

ahlinist commented on src/main/java/com/example/demo/controllers/ProductController.java:
```diff
@@ -0,0 +1,61 @@
+
+    @SneakyThrows
+    @GetMapping("/")
+    public List<? extends Product> currentUserProducts(@AuthenticationPrincipal User user, @RequestParam String productType){
```
Why ? extends instead of simply Product?

BrainDias replied:
Yes there is one with @PreAuthorize, but in this case I need authentication and user as my JPA entity at once, that's why I choose this method.

BrainDias replied:
Described the reason below

ahlinist commented on src/main/java/com/example/demo/controllers/ProductController.java:
```diff
@@ -0,0 +1,61 @@
+
+    @SneakyThrows
+    @GetMapping("/")
+    public List<? extends Product> currentUserProducts(@AuthenticationPrincipal User user, @RequestParam String productType){
```
Any chance to hide authentication from the method signatures?

BrainDias replied:
Yes there is one with @PreAuthorize, but in this case I need authentication and user as my JPA entity at once, that's why I choose this method.

ahlinist commented on src/main/java/com/example/demo/controllers/ProductController.java:
```diff
@@ -0,0 +1,61 @@
+    @SneakyThrows
+    @GetMapping("/")
+    public List<? extends Product> currentUserProducts(@AuthenticationPrincipal User user, @RequestParam String productType){
+        Stream<Product> productStream = user.getProducts().stream().filter(product -> product.getClass().getName().equals(productType));
```
Business logic should not be processed in controllers in any way, filtering out by product type is to be placed into the service.

ahlinist replied:
A single object is used at the same time as auth principal and a business object which holds child relations. 

The concerns are not well-separated.

ahlinist commented on src/main/java/com/example/demo/controllers/ProductController.java:
```diff
@@ -0,0 +1,61 @@
+    @GetMapping("/")
+    public List<? extends Product> currentUserProducts(@AuthenticationPrincipal User user, @RequestParam String productType){
+        Stream<Product> productStream = user.getProducts().stream().filter(product -> product.getClass().getName().equals(productType));
+        return userService.mapToType(productStream,productType);
```
it's considered as bad practice to pass streams to methods

service should not map to any kinds of DTOs, create a separate mapper

it's not common to pass anything to mapper's method signature, the model-to-dto mapping should be straightforward

ahlinist commented on src/main/java/com/example/demo/controllers/ProductController.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    @GetMapping("/{id}")
+    public List<? extends Product> otherUserProducts(@PathVariable Long id,@RequestParam String productType){
```
the method above is secured, and this one isn't?

BrainDias replied:
It's a method for any user to look at other user's products (if they didn't hide them), that's why.

ahlinist commented on src/main/java/com/example/demo/controllers/ProductController.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    @ResponseStatus(HttpStatus.CREATED)
+    @PutMapping("")
```
the first method has "/" in path, this one doesn't. is there any difference?

ahlinist commented on src/main/java/com/example/demo/controllers/ProductController.java:
```diff
@@ -0,0 +1,61 @@
+
+    @ResponseStatus(HttpStatus.CREATED)
+    @PutMapping("")
+    public void addProduct(@AuthenticationPrincipal User user, @RequestBody ProductDTO productDTO, @RequestParam String productType){
```
no query params should be passed in PUT request

ahlinist commented on src/main/java/com/example/demo/controllers/ProductController.java:
```diff
@@ -0,0 +1,61 @@
+    @PutMapping("")
+    public void addProduct(@AuthenticationPrincipal User user, @RequestBody ProductDTO productDTO, @RequestParam String productType){
+        Product product = null;
+        switch (productType){
```
This should be moved to the mapper. The controller just accepts DTO, maps it to a model (or entity) object, invokes the service, maps service's response back to DTO. That's it.

ahlinist commented on src/main/java/com/example/demo/controllers/ProductController.java:
```diff
@@ -0,0 +1,61 @@
+            case "pistol"-> product = mapper.productDtoToPistol(productDTO);
+            case "sniper"-> product = mapper.productDtoToSniperRifle(productDTO);
+        }
+        userService.addProduct(user,product);
```
Authentication principal is passed from controller to the service. If you need userId or something in the service - get it from security context.

BrainDias replied:
@AuthenticationPrincipal resolves security data into my entity object automatically, in addProduct() the ElementCollection of passed User entity being updated and then User being saved.

ahlinist commented on src/main/java/com/example/demo/controllers/ProductController.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    @DeleteMapping("/{index}")
```
it's more common to name those as "id", not "index", but this not a big deal anyway

ahlinist commented on src/main/java/com/example/demo/services/UserService.java:
```diff
@@ -0,0 +1,82 @@
+
+@Service
+@RequiredArgsConstructor
+public class UserService {
```
Should implement interface. And the controller should hold UserService as an interface type.

ahlinist commented on src/main/java/com/example/demo/services/UserService.java:
```diff
@@ -0,0 +1,82 @@
+
+    @Autowired
+    public static UserRepository userRepository;
+    private final DtoMapper mapper;
```
Mapper should belong to the controller. The best way is to keep separate mappers for every DTO object (e.g. ProductMapper), each of them to hold 2 methods: from DTO to Model, from Model to DTO.

BrainDias replied:
Okay, mapper now belongs to the controller only but what about separating them there is a problem since there are different classes extending Product class and if I would separate mappers I would get many mapper dependencies in controller BUT they are being used only in one context (in one method) so I created one mapper for all. Should I separate them?

ahlinist commented on src/main/java/com/example/demo/services/UserService.java:
```diff
@@ -0,0 +1,82 @@
+    public static UserRepository userRepository;
+    private final DtoMapper mapper;
+
+    public Page<User> pageUsers(Pageable pageRequest){
```
Who uses it? No one? YAGNI.

BrainDias replied:
To be used by admins/moderators in future updates.

ahlinist commented on src/main/java/com/example/demo/services/UserService.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    public void banUser(Long id) {
+        userRepository.findById(id).ifPresent(User -> {
```
variable name should start from lower case letter (see conventions)

ahlinist commented on src/main/java/com/example/demo/services/UserService.java:
```diff
@@ -0,0 +1,82 @@
+        });
+    }
+
+    public void addProduct(User user, Product product) {
```
I doubt that Product-related operations need to be performed in UserService. I'm not sure that his design is good though. I'd make Product a "first-class citizen" and set userId as a field.

ahlinist commented on src/main/java/com/example/demo/services/UserService.java:
```diff
@@ -0,0 +1,82 @@
+        List<Product> products = user.getProducts();
+        String productType = products.get(index).getClass().getName();
+        switch (productType){
+            case "Ammo" -> products.set(index,mapper.productDtoToAmmo(product));
```
white space after comma

ahlinist commented on src/main/java/com/example/demo/services/UserService.java:
```diff
@@ -0,0 +1,82 @@
+    public void updateProduct(User user, int index, ProductDTO product) {
+        List<Product> products = user.getProducts();
+        String productType = products.get(index).getClass().getName();
+        switch (productType){
```
service should not be aware of any DTOs due to law of Demeter

BrainDias replied:
Updated (not sure if it's right way to do it but idk how else to do it)

ahlinist commented on src/main/java/com/example/demo/services/UserService.java:
```diff
@@ -0,0 +1,82 @@
+        userRepository.save(user);
+    }
+
+    public List<? extends Product> otherUserProducts(Long id, String productType) {
```
why "? extends"?

BrainDias replied:
Because it returns a List of types extending Product (weapons,ammo etc.)

ahlinist commented on src/main/java/com/example/demo/services/UserService.java:
```diff
@@ -0,0 +1,82 @@
+
+    public List<? extends Product> otherUserProducts(Long id, String productType) {
+        User user = getUser(id);
+        Stream<Product> productStream = user.getProducts().stream().filter(product -> product.getClass().getName().equals(productType)&&(!product.hidden));
```
the chain is too long and unreadable, the filter predicate could be externalized as a method or something.

Also, this is a bad practice to rely on classnames is business logic.

whitespaces around &&

BrainDias replied:
I know it's not a good way to rely on class names but I really have no idea how else I can do it. Product types have different attributes, so I can't simply cut them off, and making some Product (super class) attribute which would contain the class names looks redundant while there are inbuilt Java mechanic like that.

Okay, I see one way to do it leaving only Product class - adding enum attribute to it, and depending on that attribute value getters and setters should check if an other attribute in question is allowed/disallowed to get/set for that product type. But isn't it redundant too?

ahlinist commented on src/main/java/com/example/demo/services/UserService.java:
```diff
@@ -0,0 +1,82 @@
+    public List<? extends Product> otherUserProducts(Long id, String productType) {
+        User user = getUser(id);
+        Stream<Product> productStream = user.getProducts().stream().filter(product -> product.getClass().getName().equals(productType)&&(!product.hidden));
+        return mapToType(productStream,productType);
```
whitespace

stream as an argument

ahlinist commented on src/main/java/com/example/demo/controllers/ProductController.java:
```diff
@@ -0,0 +1,61 @@
+@RequiredArgsConstructor
+@RequestMapping("/products")
+public class ProductController {
+    private final UserService userService;
```
ProductController, but no ProductService, just UserService

BrainDias replied:
Since product is just a data class and related as ManyToOne to User, can't I keep it as it is with no dedicated service just for products?

ahlinist commented on src/main/java/com/example/demo/services/UserService.java:
```diff
@@ -0,0 +1,82 @@
+
+    public static User getUser(Long id) {
+        Optional<User> optionalUser = userRepository.findById(id);
+        if(optionalUser.isEmpty()) throw new IllegalArgumentException("User doesn't exist");
```
could be done as orElseThrow

BrainDias replied:
Replaced with orElseThrow.

BrainDias committed:
1. Added UserRepository to make UserService work properly
2. Created external method for predicate in otherUserProducts() for better readability
3. Added spring data jpa to dependencies
4. Replaced NULL check -> exception throwing with orElseThrow() in getUser()

BrainDias committed:
1. Added User class to make logic work properly.
2. Updated otherUserProducts() method accordingly

BrainDias committed:
1. Externalized selecting the correct product type to mapper
2. Minor changes for cleaning code
